### DISCUSSION TITLE
Question about using vprec backend with instrument

### DISCUSSION THREAD

**Discussion by Kim1230** [verificarlo/verificarlo#326] [Q&A] [answered]
Hello, I use vprec.so backend with inst-func to profile a function. And I have some questions for understanding the output
The function is 
![image](https://user-images.githubusercontent.com/107012598/201811092-417c0741-74ea-4691-a3a5-0779e684bd6b.png)
while the input of the function are:
#define _threadargsproto_ int _iml, int _cntml_padded, double* _p, Datum* _ppvar, ThreadDatum* _thread, NrnThread* _nt, double v
And the local variables are:
#define mggate _thread[_gth]._pval[0]
#define A_AMPA _p[19*_STRIDE]
#define B_AMPA _p[20*_STRIDE]
#define A_NMDA _p[21*_STRIDE]
#define B_NMDA _p[22*_STRIDE]

And when I get the output:
> x86_64/corenrn/mod2c/DetAMPANMDA.cpp/_nrn_cur__DetAMPANMDA/_ZN10coreneuronL25_nrn_current__DetAMPANMDAEiiPdPiPNS_11ThreadDatumEPNS_9NrnThreadEdd/719/49	0	0	0	1	52	11	23	8	2	2	988200000
input:	_p	**3**	52	11	2147483647	-2147483648
input:	_thread	1	52	11	-77	13
output:	return_value	1	52	11	-1	1
output:	_p	3	52	11	2147483647	-2147483648
x86_64/corenrn/mod2c/DetAMPANMDA.cpp/_nrn_state__DetAMPANMDA/**exp**/413/91	1	0	0	1	52	11	23	8	1	1	988200000
input:	**parameter_1**	1	52	11	-1	0
output:	return_value	1	52	11	0	1

The question will be :
1.pe of input is 3, what is that meaning ?
2. And I have an input double v. But I can't see the profiling of it 
3. And _p is a list. How I can understand its maximum value(2147483647) and minimum value(-2147483648)? 
4. And whether the second function is the internal function  and what is the parameter_1 meaning?  

Hope for your response! Thank!

**Comment by pablooliveira** [comment, marked as answer]
Hi @Kim1230 

@Avaquet wrote the FuncInstrument path, I know how it works, but I am not a specialist on this part of the code.

1.  Type 3 is a pointer to a double  (double *). The types are defined in the enum inside `interflop.h` here: https://github.com/verificarlo/verificarlo/blob/3bf4a3d237295a4d0e5ee792d3b114f02dc0f68e/src/common/interflop.h#L30
2. I think the name was lost while optimizing but it seems to be called `parameter_1` here ? Only scalar arguments are recorded by FuncInstrument.
3. Even if _p is a list, I think funcInstrument treats it as a scalar pointer, so only the value _p[0] is analyzed.
4. The internal function is not captured; I think parameter_1 corresponds to v.

You seem to be doing very interesting stuff with verificarlo; if you get interesting insights, do not hesitate to report back here ! I would be happy to see where your efforts lead :-)
